src/evaluate/evaluator/question_answering.py

```diff
@@ -0,0 +1,191 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Callable, Dict, Optional, Tuple, Union
+
+# Lint as: python3
+from datasets import Dataset, load_dataset
+
+
+try:
+    from transformers import Pipeline, PreTrainedModel, PreTrainedTokenizer, TFPreTrainedModel
+
+    TRANSFORMERS_AVAILABLE = True
+except ImportError:
+    TRANSFORMERS_AVAILABLE = False
+
+from typing_extensions import Literal
+
+from ..module import EvaluationModule
+from ..utils.logging import get_logger
+from .base import Evaluator
+
+
+logger = get_logger(__name__)
+
+
+class QuestionAnsweringEvaluator(Evaluator):
+    """
+    Question answering evaluator.
+    This question answering evaluator can currently be loaded from [`evaluator`] using the default task name
+    `question-answering`.
+    Methods in this class assume a data format compatible with the [`QuestionAnsweringPipeline`].
+    """
+
+    def __init__(self, task="question-answering", default_metric_name=None):
+        super().__init__(task, default_metric_name=default_metric_name)
+
+    def prepare_data(
+        self,
+        data: Union[str, Dataset],
+        question_column: str,
+        context_column: str,
+        id_column: str,
+        label_column: str,
+    ):
+        """Prepare data."""
+        if data is None:
+            raise ValueError(
+                "Please specify a valid `data` object - either a `str` with a name or a `Dataset` object."
+            )
+        data = load_dataset(data) if isinstance(data, str) else data
+        if question_column not in data.column_names:
+            raise ValueError(
+                f"Invalid `question_column` {question_column} specified. The dataset contains the following columns: {data.column_names}."
+            )
+        if context_column not in data.column_names:
+            raise ValueError(
+                f"Invalid `context_column` {context_column} specified. The dataset contains the following columns: {data.column_names}."
+            )
+        if id_column not in data.column_names:
+            raise ValueError(
+                f"Invalid `id_column` {id_column} specified. The dataset contains the following columns: {data.column_names}."
+            )
+        if label_column not in data.column_names:
+            raise ValueError(
+                f"Invalid `label_column` {label_column} specified. The dataset contains the following columns: {data.column_names}."
+            )
+
+        return data
```

Comment:
```
        data = super().prepare_data(data=data, input_column=question_column, label_column=label_column)

        if context_column not in data.column_names:
            raise ValueError(
                f"Invalid `context_column` {context_column} specified. The dataset contains the following columns {data.column_names}."
            )
        if id_column not in data.column_names:
            raise ValueError(
                f"Invalid `id_column` {id_column} specified. The dataset contains the following columns: {data.column_names}."
            )

        return data
```

Reply:
We should really give a dict there and loop over key, values.